models/batch.js

```diff
@@ -48,17 +50,25 @@ Batch.prototype.lock = function(userId, query, callback) {
     read: false
   });
 
-  Report
-    .find(filter)
-    .sort({ storedAt: -1 })
-    .limit(ITEMS_PER_BATCH)
-    .exec(function(err, reports) {
-      if (err) return callback(err);
-      var ids = _.map(reports, '_id');
-      var update = { checkedOutBy: userId, checkedOutAt: new Date() };
-      Report.update({ _id: { $in: ids } }, update, { multi: true }, callback);
-    }
-  );
+  lock.writeLock(function(release) {
```

Comment:
Does `findAndModify` in mongo not work for this? I would expect there would be a way to do this without resorting to external locks. It's generally going to be cleaner to keep it within mongo too, since, for example, `rwlock` will probably only be able to lock within one process.

Reply:
I came to this bug happily thinking that what you said was true. Alas, ```findAndModify``` only works with single documents. ```update``` does work with several documents but does not ```sort``` nor ```limit```. The ```batch.lock``` and ```batch.load```  is a mechanism around those limitations.

There are other locking mechanism that work across processes, e.g., https://github.com/xhawk18/key_mutex , however our current ```cluster_app``` is not working.

Reply:
Darn, okay. Would `$isolated` work here?

Reply:
I used it and did not solve anything by itself. May help in other extreme cases, however, I removed it as I was not sure about the performance penalty it will add in MongoDB.